firmware/library/L1_Drivers/onboardled.hpp

```diff
@@ -18,3 +20,81 @@ class OnBoardLedInterface {
            virtual void LedSetAll(uint8_t value)                     = 0;
            virtual uint8_t LedGetValues(void)                        = 0;
 };
+
+class OnBoardLed: public Gpio, public OnBoardLedInterface
+{
+ public:
+  static constexpr uint8_t Led0Position = 0;
+  static constexpr uint8_t Led1Position = 1;
+  static constexpr uint8_t Led2Position = 2;
+  static constexpr uint8_t Led3Position = 3;
+
+  // Initialization involes setting the appropriate pins to be GPIO, and
+  // then outputs. Once they are configured as such, they can be maniputlated
+  // as expected.
+  void Initialize(void) override
+    {
+    // This block uses the predefined GPIO class.
+    Gpio Led0(1, 1);
+    Gpio Led1(1, 8);
+    Gpio Led2(1, 10);
+    Gpio Led3(1, 15);
+    // This block will make sure all the GPIOs that were initialized in the
+    // last block will be set as outputs.
+    Led0.SetAsOutput();
+    Led1.SetAsOutput();
+    Led2.SetAsOutput();
+    Led3.SetAsOutput();
+    // This block turns off all the LEDs by default, to save power.
+    led_states = 0xFF;
+    LedSetAll(led_states);
+  }
+
+  void LEDOn(uint8_t led_number) override
+  {
+    led_states = (led_states & ~(1 << led_number));
+    LedSetAll(led_states);
+  }
+
+  void LEDOff(uint8_t led_number) override
+  {
+    led_states = (led_states | (1 << led_number));
+    LedSetAll(led_states);
+  }
+
+  void LedSet(uint8_t led_number, bool on) override
+  {
+     on ? LedOn(led_number); : LedOff(led_number);
+  }
+
+  void LedToggle(uint8_t led_number) override
+  {
+    led_states = (led_states ^ (1 << led_number));
+    LedSetAll(led_states);
+  }
+
+  void LedSetAll(uint8_t value)
+  {
+  led_states = value;
+  if (led_states&(1 << Led0Position))
+  {
+  Led0.SetHigh();
+  }
+  else{
+  Led0.SetLow();
+  }
+
+
+  (led_states&(1 << Led1Position))? Led1.SetHigh(); : Led1.SetLow();
+  (led_states&(1 << Led2Position))? Led2.SetHigh(); : Led2.SetLow();
+  (led_states&(1 << Led3Position))? Led3.SetHigh(); : Led3.SetLow();
```

Comment:
Why are these ternary expression when the above is a full if-else block. Would prefer if-else over ternary function calls. But even better would be a solution that doesn't require conditionals.